feat: Add docs to install Nginx Gateway Fabric with N+ (#1524)
nginxinc/nginx-gateway-fabric: site/content/installation/ngf-images/pulling-ngf-image.md

---

> title: Pushing an NGINX Plus image for NGINX Gateway Fabric to a private registry

Reviewer: Looking at the rendered docs, these two titles for the new docs are a bit long. I think for the titles let's just say `NGINX Plus image` instead of `NGINX Plus image for NGINX Gateway Fabric`. The description can still include it, but the titles are a bit long.

---
title: Push NGINX Plus Image to Private Registry
description: "Learn how to pull an F5 NGINX Plus Gateway Fabric image from the official F5 Docker registry and upload it to your private registry. This guide covers the prerequisites, image tagging, and troubleshooting steps."
weight: 100
doctypes: ["install"]
toc: true
---

> {{<custom-styles>}}
> 
> 

Reviewer 2: ```suggestion
This guide covers the prerequisites, image tagging, and troubleshooting steps on Linux. Alternative instructions for Windows and macOS operating systems are provided where relevant.
```

## Before you begin

Before you start, you'll need these installed on your machine:

- [Docker v18.09 or higher](https://docs.docker.com/engine/release-notes/18.09/).

> - An NGINX Gateway Fabric subscription. Download both the certificate (*nginx-repo.crt*) and key (*nginx-repo.key*) from the [MyF5 Customer Portal](https://my.f5.com). Keep in mind that an NGINX Plus certificate and key won't work for for the steps in this guide.

Reviewer 2: ```suggestion
- The certificate (**nginx-repo.crt**) and key (**nginx-repo.key**) for an NGINX Gateway Fabric Subscription, obtainable from [MyF5l](https://my.f5.com) An NGINX Plus certificate and key will not work.
```

> ## Set up Docker for F5 Container Registry

Reviewer 2: ```suggestion
## Configuring Docker for the F5 Container Registry
```

> Start by setting up Docker to communicate with the F5 Container Registry located at `private-registry.nginx.com`. If you're using Linux, follow these steps to create a directory and add your certificate and key:

Reviewer 2: ```suggestion
To configure Docker to communicate with the F5 Container Registry, first create a folder containing your certificate and key files:
```

```shell
mkdir -p /etc/docker/certs.d/private-registry.nginx.com
cp <path-to-your-nginx-repo.crt> /etc/docker/certs.d/private-registry.nginx.com/client.cert
cp <path-to-your-nginx-repo.key> /etc/docker/certs.d/private-registry.nginx.com/client.key
```

> The steps provided are for Linux. For Mac or Windows, consult the [Docker for Mac](https://docs.docker.com/desktop/faqs/general/#how-do-i-add-custom-ca-certificates) or [Docker for Windows](https://docs.docker.com/desktop/faqs/windowsfaqs/#how-do-i-add-custom-ca-certificates) documentation. For more details on Docker Engine security, you can refer to the [Docker Engine Security documentation](https://docs.docker.com/engine/security/).

Reviewer 2: ```suggestion
If you are not using a Linux operating system, read the [Docker for Windows](https://docs.docker.com/desktop/faqs/windowsfaqs/#how-do-i-add-custom-ca-certificates) or [Docker for Mac](https://docs.docker.com/desktop/faqs/macfaqs/#add-custom-ca-certificates-server-side) instructions. For more details on Docker Engine security, you can refer to the [Docker Engine Security documentation](https://docs.docker.com/engine/security/).
```

> ## Pull the image

Reviewer 2: ```suggestion
## Pulling the image
```

> Next, pull the image you need from `private-registry.nginx.com`. To find the correct image, consult the [Technical Specifications](https://github.com/nginxinc/nginx-gateway-fabric#technical-specifications).

Reviewer 2: ```suggestion
Once configured, you can now pull images from `private-registry.nginx.com`. To find your desired image, read the [Technical Specifications](https://github.com/nginxinc/nginx-gateway-fabric#technical-specifications).
```

> To pull an image, follow these steps. Replace `<version-tag>` with the specific version you need, for example, `1.2.0`.

Reviewer 2: ```suggestion
Run this command step to pull an image, replacing `<version-tag>` with the specific version you need, such as `1.2.0`.
```

> 
> - For NGINX Plus Gateway Fabric, run:
> 

Reviewer 2: ```suggestion
```

  ```shell

>   docker pull private-registry.nginx.com/nginx-gateway-fabric/nginx-plus:<version-tag>

Reviewer: ```suggestion
  docker pull private-registry.nginx.com/nginx-gateway-fabric/nginx-plus:1.2.0
```
Since it's the example, I'd include the version.

  ```

> You can use the Docker registry API to list the available image tags by running the following commands. Replace `<path-to-client.key>` with the location of your client key and `<path-to-client.cert>` with the location of your client certificate. The `jq` command is used to format the JSON output for easier reading.

Reviewer 2: ```suggestion
You can use the Docker registry API to list available image tags using your client certificate and key. The `jq` command is used to format the JSON output for easier reading.
```

```json

> $ curl https://private-registry.nginx.com/nginx-gateway-fabric/nginx-plus/tags/list --key <path-to-client.key> --cert <path-to-client.cert> | jq

Reviewer: Include this command in a separate `shell` block, and the output can be in a `json` block. Makes it easier for a user to just copy the command.

> $ curl https://private-registry.nginx.com/nginx-gateway-fabric/nginx-plus/tags/list --key <path-to-client.key> --cert <path-to-client.cert> | jq

Reviewer 2: ```suggestion
curl https://private-registry.nginx.com/nginx-gateway-fabric/nginx-plus/tags/list --key <path-to-client.key> --cert <path-to-client.cert> | jq
```

{
  "name": "nginx-gateway-fabric/nginx-plus",
  "tags": [
    "edge"
  ]
}
```

> 

Reviewer 2: ```suggestion
```

> ## Push to your private registry

Reviewer 2: ```suggestion
## Pushing an image to a private repository
```

> After pulling the image, tag it and upload it to your private registry.

Reviewer 2: ```suggestion
Once you have pulled an image, you can tag it and push it to a private registry.
```

> 1. Log in to your private registry:

Reviewer 2: ```suggestion
1. Log into your private registry:
```

   ```shell
   docker login <my-docker-registry>
   ```

> 1. Tag and push the image. Replace `<my-docker-registry>` with your registry's path and `<version-tag>` with the version you're using, for example `1.2.0`:
> 
>    - For NGINX Plus Gateway Fabric, run:
> 
>       ```shell
>       docker tag private-registry.nginx.com/nginx-gateway-fabric/nginx-plus:<version-tag> <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>
>       docker push <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>
>       ```
> 
> 

Reviewer 2: ```suggestion
1. Tag the image, replacing `<my-docker-registry>` with your registry's path and `<version-tag>` with the version you're using:

    ```shell
        docker tag private-registry.nginx.com/nginx-gateway-fabric/nginx-plus:<version-tag> <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>
   
1. Push the image:

    ```shell
        docker push <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>
    ```
```

## Troubleshooting

If you encounter issues while following this guide, here are solutions to common problems:

> - **Certificate errors**
>   - **Likely Cause**: Incorrect certificate or key location, or using an NGINX Plus certificate.
>   - **Solution**: Verify you have the correct NGINX Gateway Fabric certificate and key. Place them in the correct directory and ensure the certificate has a *.cert* extension.
> 
> - **Docker version compatibility**
>   - **Likely Cause**: Outdated Docker version.
>   - **Solution**: Make sure you're running [Docker v18.09 or higher](https://docs.docker.com/engine/release-notes/18.09/). Upgrade if necessary.
> 
> - **Can't pull the image**
>   - **Likely Cause**: Mismatched image name or tag.
>   - **Solution**: Double-check the image name and tag against the [Tech Specs guide](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CHANGELOG.md).
> 
> - **Failed to push to private registry**
>   - **Likely Cause**: Not logged into your private registry or incorrect image tagging.
>   - **Solution**: Verify login status and correct image tagging before pushing. Consult the [Docker documentation](https://docs.docker.com/docker-hub/repos/) for more details.

Reviewer 2: ```suggestion
- **Certificate errors**:
  - *Likely cause*: Incorrect certificate or key location, or using an NGINX Plus certificate.
  - *Solution*: Check you have the correct NGINX Gateway Fabric certificate and key, their files are named correctly, and they are in the correct directory.

- **Docker version compatibility**
  - *Likely cause*: Outdated Docker version.
  - *Solution*: Make sure you're running [Docker v18.09 or higher](https://docs.docker.com/engine/release-notes/18.09/), and upgrade if necessary.

- **Can't pull the image**
  - *Likely cause*: Mismatched image name or tag.
  - *Solution*: Compare the image name and tag to the [Technical Specifications table](https://github.com/nginxinc/nginx-gateway-fabric?tab=readme-ov-file#technical-specifications).

- **Failed to push to private registry**
  - *Likely cause*: Not logged into your private registry or incorrect image tagging.
  - *Solution*: Verify your login status and correct the image tag before pushing. Read the [Docker documentation](https://docs.docker.com/docker-hub/repos/) for more guidance.
```

## Alternative installation options

> You can also get the NGINX Plus image using the following alternate methods:

Reviewer 2: ```suggestion
There are alternative ways to get an NGINX Plus Gateway Fabric Image:
```

- [Install using a JWT token in a Docker Config Secret]({{< relref "jwt-token-docker-secret" >}}).

> - [Build the Gateway Fabric image](https://docs.nginx.com/nginx-gateway-fabric/installation/ngf-images/building-the-images/) using the source code from the GitHub repository and your NGINX Plus subscription certificate and key.

Reviewer: same comment about using `relref` instead of the link.

> - [Build the Gateway Fabric image](https://docs.nginx.com/nginx-gateway-fabric/installation/ngf-images/building-the-images/) using the source code from the GitHub repository and your NGINX Plus subscription certificate and key.

Reviewer: ```suggestion
- [Build the NGINX Plus image](https://docs.nginx.com/nginx-gateway-fabric/installation/ngf-images/building-the-images/) using the source code from the GitHub repository and your NGINX Plus subscription certificate and key.
```

> - [Install using a JWT token in a Docker Config Secret]({{< relref "jwt-token-docker-secret" >}}).
> - [Build the Gateway Fabric image](https://docs.nginx.com/nginx-gateway-fabric/installation/ngf-images/building-the-images/) using the source code from the GitHub repository and your NGINX Plus subscription certificate and key.

Reviewer 2: ```suggestion
- [Getting the NGINX Plus Gateway Fabric Image using JWT]({{< relref "/installation/ngf-images/jwt-token-docker-secret.md" >}}).
```

---
title: Pushing an NGINX Plus Gateway Fabric image to a private registry

> description: "This document describes how to pull a NGINX Plus Gateway Fabric image from the official F5 Docker registry and upload it to your private registry."

Reviewer: ```suggestion
description: "This document describes how to pull an NGINX Plus image for NGINX Gateway Fabric from the official F5 Docker registry and upload it to your private registry."
```

> weight: 100

Reviewer: Can we change this weight to 300? Also, let's change the weight of `installing-ngf/_index.md` to 100, and `running-on-kind.md` to 400.

I'd like the order on the docs page to look like `Install`, `Expose`, `Images`, then `deploy on kind`.

Reviewer: These aren't showing up in this order yet. Be sure to check out the netlify preview to ensure the order that the pages are displayed.

Reviewer: This order is still wrong. Please check out the deploy preview. Should be `Install`, `Expose`, `Images`, then `deploy on kind`

doctypes: ["install"]
toc: true
---

{{<custom-styles>}}

## Before you begin

Before you start, you'll need these installed on your machine:

- [Docker v18.09 or higher](https://docs.docker.com/engine/release-notes/18.09/).

> - The certificate (**nginx-repo.crt**) and key (**nginx-repo.key**) for an NGINX Gateway Fabric Subscription, obtainable from [MyF5l](https://my.f5.com) An NGINX Plus certificate and key will not work.

Reviewer: ```suggestion
- The certificate (**nginx-repo.crt**) and key (**nginx-repo.key**) for a Connectivity Stack for Kubernetes subscription, obtainable from [MyF5l](https://my.f5.com) An NGINX Plus certificate and key will not work.
```

## Pulling the image

Once configured, you can now pull images from `private-registry.nginx.com`. To find your desired image, read the [Technical Specifications](https://github.com/nginxinc/nginx-gateway-fabric#technical-specifications).

Run this command step to pull an image, replacing `<version-tag>` with the specific version you need, such as `1.2.0`.

  ```shell
  docker pull private-registry.nginx.com/nginx-gateway-fabric/nginx-plus:1.2.0
  ```

You can use the Docker registry API to list available image tags using your client certificate and key. The `jq` command is used to format the JSON output for easier reading.

```shell
curl https://private-registry.nginx.com/nginx-gateway-fabric/nginx-plus/tags/list --key <path-to-client.key> --cert <path-to-client.cert> | jq
```

```json
{
  "name": "nginx-gateway-fabric/nginx-plus",
  "tags": [
    "edge"
  ]
}
```

Once you have pulled an image, you can tag it and push it to a private registry.

1. Log into your private registry:

   ```shell
   docker login <my-docker-registry>
   ```

1. Tag the image, replacing `<my-docker-registry>` with your registry's path and `<version-tag>` with the version you're using:

>    - For NGINX Plus Gateway Fabric, run:

Reviewer: Since this doc is only for NGINX Plus, I don't think this line is needed.

    ```shell
    docker tag private-registry.nginx.com/nginx-gateway-fabric/nginx-plus:<version-tag> <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>
    docker push <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>
    ```

    ```shell
    docker push <my-docker-registry>/nginx-gateway-fabric/nginx-plus:<version-tag>

>     ```

Reviewer: This is duplicated from the lines before, so can be removed.

---

> title: Getting the F5 Registry NGINX Plus Gateway Fabric Image

Reviewer: ```suggestion
title: Push NGINX Plus Image to Private Registry
```

maybe doc-ops can help phrase this better for a page title. Also, NGINX Plus Gateway Fabric is not the product name, so I think we should probably just say NGINX Plus image throughout the docs.

Author: updating to your suggestion for now

description: "Learn how to pull an F5 NGINX Plus Gateway Fabric image from the official F5 Docker registry and upload it to your private registry. This guide covers the prerequisites, image tagging, and troubleshooting steps."
weight: 100
doctypes: ["install"]
toc: true
---

{{<custom-styles>}}

---

## Before you begin

Before you start, you'll need these installed on your machine:

- [Docker v18.09 or higher](https://docs.docker.com/engine/release-notes/18.09/).
- An NGINX Gateway Fabric subscription. Download both the certificate (*nginx-repo.crt*) and key (*nginx-repo.key*) from the [MyF5 Customer Portal](https://my.f5.com). Keep in mind that an NGINX Plus certificate and key won't work for for the steps in this guide.

---

## Set up Docker for F5 Container Registry

> Start by setting up Docker to commungfate with the F5 Container Registry located at `private-registry.nginx.com`. If you're using Linux, follow these steps to create a directory and add your certificate and key:

Author: I got this link from Luca for accessing private registry's @lucacome.

> Start by setting up Docker to commungfate with the F5 Container Registry located at `private-registry.nginx.com`. If you're using Linux, follow these steps to create a directory and add your certificate and key:

Author: Luca mentioned this would be the correct link to refer to Nginx Plus images.

```shell
mkdir -p /etc/docker/certs.d/private-registry.nginx.com
cp <path-to-your-nginx-repo.crt> /etc/docker/certs.d/private-registry.nginx.com/client.cert
cp <path-to-your-nginx-repo.key> /etc/docker/certs.d/private-registry.nginx.com/client.key
```

The steps provided are for Linux. For Mac or Windows, consult the [Docker for Mac](https://docs.docker.com/desktop/faqs/general/#how-do-i-add-custom-ca-certificates) or [Docker for Windows](https://docs.docker.com/desktop/faqs/windowsfaqs/#how-do-i-add-custom-ca-certificates) documentation. For more details on Docker Engine security, you can refer to the [Docker Engine Security documentation](https://docs.docker.com/engine/security/).

---

## Pull the image

> Next, pull the image you need from `private-registry.nginx.com`. To find the correct image, consult the [Tech Specs guide](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CHANGELOG.md).

Reviewer: We should probably link to releases, not the CHANGELOG, right?

Reviewer 2: I think the scope of this confusion is actually wider than it initially looks.

The repo README has a [technical specifications section}(https://github.com/nginxinc/nginx-gateway-fabric#technical-specifications), but pulling back even farther, there doesn't appear to be a good reference table for all available images.

For comparison, we currently direct people to look at the README on an otherwise sparse page on the documentation site:

https://docs.nginx.com/nginx-gateway-fabric/reference/technical-specifications/

The equivalent NGINX Ingress Controller page, however, lists the actual image names in tables based on registry.

https://docs.nginx.com/nginx-ingress-controller/technical-specifications/#supported-docker-images

The quick fix for this is to change the link to the README section above (I would also change the text to "consult the Technical Specifications" to match the heading", but tossing a ticket in the backlog to add a proper reference table for images would also great for the user experience.

Author: udpating

Author: updating

> Next, pull the image you need from `private-registry.nginx.com`. To find the correct image, consult the [Tech Specs guide](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CHANGELOG.md).

Reviewer 2: ```suggestion
Next, pull the image you need from `private-registry.nginx.com`. To find the correct image, consult the [Technical Specifications][(https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CHANGELOG.md](https://github.com/nginxinc/nginx-gateway-fabric#technical-specifications.
```
This edit suggestion follows the conversation Saylor started in another thread.

Author: updating

> To pull an image, follow these steps. Replace `<version-tag>` with the specific version you need, for example, `3.4.2`.

Reviewer: Let's use 1.2.0 throughout, since that's a real version.

> To pull an image, follow these steps. Replace `<version-tag>` with the specific version you need, for example, `3.4.2`.

Reviewer 2: Using a real version also means that a mechanism for version bumping in the future could allow the given example to always show the latest tag in documentation.

> To pull an image, follow these steps. Replace `<version-tag>` with the specific version you need, for example, `3.4.2`.

Author: updating

- **Certificate errors**
  - **Likely Cause**: Incorrect certificate or key location, or using an NGINX Plus certificate.
  - **Solution**: Verify you have the correct NGINX Gateway Fabric certificate and key. Place them in the correct directory and ensure the certificate has a *.cert* extension.

- **Docker version compatibility**
  - **Likely Cause**: Outdated Docker version.
  - **Solution**: Make sure you're running [Docker v18.09 or higher](https://docs.docker.com/engine/release-notes/18.09/). Upgrade if necessary.

- **Can't pull the image**
  - **Likely Cause**: Mismatched image name or tag.
  - **Solution**: Double-check the image name and tag against the [Tech Specs guide](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CHANGELOG.md).

- **Failed to push to private registry**
  - **Likely Cause**: Not logged into your private registry or incorrect image tagging.
  - **Solution**: Verify login status and correct image tagging before pushing. Consult the [Docker documentation](https://docs.docker.com/docker-hub/repos/) for more details.

---

## Alternative installation options

> You can also get the NGINX Gateway Fabric image using the following alternate methods:

Reviewer: ```suggestion
You can also get the NGINX Plus image using the following alternate methods:
```
